src/auth_v3.js

```diff
@@ -148,53 +135,139 @@ export function getAuthCodeURL (client, scopes = []) {
 // This method extracts the access code and state from the given URL. By
 // default it uses window.location.href. Also, it checks the given state with
 // the one specified in the URL query parameter to prevent CSRF attacks.
-export function getAccessToken (client, state, pageURL = '') {
-  if (pageURL === '' && typeof document !== 'undefined') {
-    pageURL = window.location.href
+export function getAccessToken (cozy, client, state, pageURL = '') {
+  if (!state) {
+    return Promise.reject(new Error('Missing state value'))
+  }
+  const queries = getGrantCodeFromPageURL(pageURL)
+  if (queries === null) {
+    return Promise.reject(new Error('Missing states from current URL'))
   }
-  const queries = decodeQuery(pageURL)
   if (state !== queries['state']) {
     return Promise.reject(new Error('Given state does not match url query state'))
   }
-  return retrieveToken(client, {
+  return retrieveToken(cozy, client, null, {
     'grant_type': 'authorization_code',
     'code': queries['access_code']
   })
 }
 
 // refreshToken perform a request on the access_token entrypoint with the
 // refresh_token grant type in order to refresh the given token.
-export function refreshToken (client, token) {
-  return retrieveToken(client, {
+export function refreshToken (cozy, client, token) {
+  return retrieveToken(cozy, client, token, {
     'grant_type': 'refresh_token',
     'code': token.refreshToken
   })
 }
 
+// oauthFlow perform the stateful registration and access granting of an OAuth
+// client.
+export function oauthFlow (cozy, storage, pageURL, createClient, onRegistered) {
```

Comment:
It's hard to read this function. Maybe using `await` can help to make it more straightforward